### PR TITLE
Add iml files to gitignore

### PR DESCRIPTION
**Pull request by nikhilnanivadekar**

### CODE REVIEW

**Comment by nikhilnanivadekar** [comment]
The gitignore is mainly from feedback by contributors. Every IntelliJ release leads to the iml files being churned. For the project and to have it work correctly, I do not think we need to have it in there. This makes rebasing, merging, etc. painful as we have to always stash these files first.
@donraab @mohrezaei @motlin @cguntur-bnym

File: .gitignore [hunk 1]
```diff
@@ -75,3 +76,18 @@ fabric.properties
 .project
 .settings/
 .classpath
+
+#IntelliJ Project files
```
Comment (prathasirisha) [inline .gitignore:80]: Missing this module's iml file. 
acceptance-tests

File: .gitignore [hunk 2]
```diff
@@ -21,6 +21,7 @@ test-coverage/**
 .idea/**/tasks.xml
 .idea/**/usage.statistics.xml
 .idea/**/shelf
+.idea/*.xml
```
Comment (motlin) [inline .gitignore:24]: This is xml rather than iml and includes many files which are checked in. This line needs to be removed.

File: .gitignore [hunk 3]
```diff
@@ -75,3 +76,18 @@ fabric.properties
...
+eclipse-collections/*.iml
+eclipse-collections-testutils/*.iml
+eclipse-collections-forkjoin/*.iml
+unit-tests/*.iml
+serialization-tests/*.iml
+jcstress-tests/*.iml
+junit-trait-runner/*.iml
+unit-tests-java8/*.iml
+test-coverage/*.iml
+p2-repository/*.iml
```
Comment (motlin) [inline .gitignore:93]: Replace this whole section with `*.iml`. In other words, get rid of the directory prefixes. No need for globs that only match one file.

**Comment by motlin** [comment]
`git rm -f '*.iml'`

**Comment by motlin** [comment]
Once we remove the iml files from version control, it will break all the TeamCity builds. But leaving them in version control and also ignoring them - I don't think that will have the desired effect but you can develop that way and let us know if it works.

**Comment by motlin** [comment]
My version of the same idea: #851

File: .gitignore (hunk 3 above)
Comment (nikhilnanivadekar) [reply .gitignore:93]: The global *.iml doesn’t seem to work. I have tried it before.

**Comment by nikhilnanivadekar** [comment]
No need to remove the iml files, we just don’t need to track them

File: .gitignore (hunk 2 above)
Comment (nikhilnanivadekar) [reply .gitignore:24]: This is not a mistake it’s on purpose. There are xml files which get added due to plugins or new versions for example encodings.xml, hydra.xml. These files should never be committed but cause issues during rebasing. So, having this line is correct

Comment (nikhilnanivadekar) [reply .gitignore:24]: This is not a mistake it’s on purpose. There are xml files which get added due to plugins or new versions for example encodings.xml, hydra.xml. These files should never be committed but cause issues during rebasing. So, having this line is correct

**Comment by motlin** [comment]
> No need to remove the iml files, we just don’t need to track them

Why do you think that works? I'll show you why I saw this doesn't work.

```
$ git checkout nikhil/gitIgnore -f
HEAD is now at 8e05c4ed Add iml files to gitignore as they are a major source of annoyance and inconvenience to contributors.
$ grep iml .gitignore 
eclipse-collections-code-generator/*.iml
eclipse-collections-code-generator-maven-plugin/*.iml
eclipse-collections-api/*.iml
eclipse-collections/*.iml
eclipse-collections-testutils/*.iml
eclipse-collections-forkjoin/*.iml
unit-tests/*.iml
serialization-tests/*.iml
jcstress-tests/*.iml
junit-trait-runner/*.iml
unit-tests-java8/*.iml
test-coverage/*.iml
p2-repository/*.iml
$ echo "some deliberate change" >> eclipse-collections/eclipse-collections.iml 
$ git status
HEAD detached at nikhil/gitIgnore
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   eclipse-collections/eclipse-collections.iml

no changes added to commit (use "git add" and/or "git commit -a")
```

File: .gitignore (hunk 2 above)
Comment (motlin) [reply .gitignore:24]: It's too broad. There are 2 or 3 xml files which can be ignored and a bunch which must be committed. Unfortunately, encodings.xml is one that is probably required by TeamCity. I'll test that after I test removing the imls.

Comment (nikhilnanivadekar) [reply .gitignore:24]: Sorry for being dense here. In none of the other projects I have worked with to get things to work correctly, was there a necessity for the xml files. If we want an off cycle change we can always work around it. I would rather be too broad than too restrictive

Comment (nikhilnanivadekar) [reply .gitignore:24]: Sorry for being dense here. In none of the other projects I have worked with to get things to work correctly, was there a necessity for the xml files. If we want an off cycle change we can always work around it. I would rather be too broad than too restrictive

**Comment by nikhilnanivadekar** [comment]
I am confused, doesn’t it say no changes staged for commit? I tried it in my local to keep the iml files and use different versions of intellij. The iml files change but are never shown in git status.

**Comment by donraab** [comment]
> Changes not staged for commit:

This is the desired behavior I thought?

**Comment by donraab** [comment]
I think Craig is correct... we should see the files in this section right?

> Untracked files:

**Comment by nikhilnanivadekar** [comment]
No they shouldn’t otherwise all files under target should also show up each time we build. Or it’s possible I am missing something

**Comment by nikhilnanivadekar** [comment]
handled in #851

**Pull request closed without merging**